### PR TITLE
Add nn::functional::normalize() to C++ Frontend

### PR DESCRIPTION
**Pull request by gtamba**
Addresses #27048 

PR Summary:

Files Added:

_torch/csrc/api/include/torch/nn/options/normalization.h
torch/csrc/api/include/torch/nn/functional/normalization.h_

Files Modified:

_test/cpp/api/functional.cpp
torch/csrc/api/include/torch/nn/functional.h_

---

@yf225 : I couldn't find a C++ equivalent of gradcheck(), is there such a function or is it sufficient to call .backward() in the test body? I don't think any solutions are checked for the Python tests.

### CODE REVIEW

**Comment by gtamba** [comment]
@yf225 : This might be a silly question but do I need to do anything additional to get the new files included in the build targets? I couldn't find hints from other pull requests unless they were .cpp modules but now trying to build the test binary gives me the following error 

```c
error: use of undeclared identifier 'NormalizeOptions'
    torch::Tensor norm = F::normalize(input, NormalizeOptions().p(1).dim(-1));
                                             ^
error: no member named 'normalize' in namespace 'torch::nn::functional'; did you mean 'c10::attr::normalized'?
    torch::Tensor norm = F::normalize(input, NormalizeOptions().p(1).dim(-1));
```

File: torch/csrc/api/include/torch/nn/options/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/normalization.h]: nit: would be great to write:
```suggestion
  TORCH_ARG(double, p) = 2.0;
```
to better indicate that it's a double value.

File: torch/csrc/api/include/torch/nn/options/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/normalization.h]: nit: it seems that we are missing a new line at the end of this file.

File: torch/csrc/api/include/torch/nn/options/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/normalization.h]: nit:
```suggestion
/// Options for the `normalize` functional.
```

File: torch/csrc/api/include/torch/nn/functional/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: nit: it would be great to remove the extra line here.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: It would be great to have specific values for `input` (e.g. `torch::arange(48).reshape({1, 3, 4, 4})`), and then check that the value of `norm` is what we expect it should be.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: It would be great to assert that `input.grad().numel()` is as expected, to show that the backward pass is successful.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: It would be great to have specific values for `input` (e.g. `torch::arange(48).reshape({1, 3, 4, 4})`), and then check that the value of `output` is what we expect it should be.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: We can remove this test case if you'd like, since the backward pass is already tested in the first test case.

**Review by yf225 — COMMENTED** [review]
@gtamba Thanks a lot for the great work! My sincere apologies for the delay in reviewing as I was out of office last week. I left some minor comments in the PR, and we should be able to merge it very soon :D

File: torch/csrc/api/include/torch/nn/options/normalization.h
Comment (gtamba) [reply torch/csrc/api/include/torch/nn/options/normalization.h]: Makes perfect sense this is a function not a class

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: Yes this was my first intuition as well but I decided to follow the python tests as is. Will make the necessary changes and maybe in the python tests in a future pr

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: Sure that does seem tidier, it was present in the python tests so I assume it was testing norm for a size{1} tensor

**Comment by gtamba** [comment]
@yf225  : it seems that the bot has done a merge instead of a rebase, should I reset the head to the commit before the merge, add my changes and then rebase to hide the merge commit or is that all right?

**Comment by yf225** [comment]
@gtamba : it's totally alright and I think we should work on top of the merge commit. At merge time, all commits in this PR will automatically be squashed into one commit, so we won't need to worry about the number of commits we make to the PR :D

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: I reduced the number of elements to 10 so that a readable tensor could be pasted into the test code. I can't think of another way of "generating" the solution on the fly without using the same steps used in the body of `F::normalize()` so I pre-calculated the solution using the Python API, is that considered acceptable?

My local source is still building, let's see if CI beats me to it :D

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: thanks this was the missing key and was confusing me quite a bit because input.grad() is none if you apply a reshape operation and override the identifier that holds the original unreshaped variable. arange and linspace don't have ways to specify size into their factory functions, I'm still now sure how this works in C++

My understanding as of now 

```python
input = torch.arange(10, dtype=float, requires_grad=True).reshape((1,2,5))
loss = input.norm().sum()
loss.backward()

input.grad
=> None
```

I suspect this happens because even though autograd is tracking the original variable the identifier to it lost in the reshape operation which does not have a derivative (I believe?)

File: test/cpp/api/functional.cpp
Comment (yf225) [reply test/cpp/api/functional.cpp]: Yes using pre-calculated solution from Python API is awesome. Thanks a lot for the changes! :D

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: nit: we are missing `;` at the end of this line.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: nit: we are missing `;` at the end of this line.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: nit: we are missing `;` at the end of this line.

File: torch/csrc/api/include/torch/nn/functional.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional.h]: nit: missing new line here

File: torch/csrc/api/include/torch/nn/functional/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: nit: to conform to the style of the code base, we should write
```
}
else {
```
as
```
} else {
```

File: test/cpp/api/functional.cpp
Comment (yf225) [reply test/cpp/api/functional.cpp]: I think your assessment is correct, and we can use `input = torch::tensor({{{0, 1, 2, 3, 4}, {5, 6, 7, 8, 9}}}, torch::dtype(torch::kFloat).requires_grad(true))` as tensor to get around this issue. :D

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: this is an embarrassing miss tbh :O

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: That is much neater I agree xD

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: Not sure what happened here, I think I fixed the semicolon on my local and forgot the to push the changes before building my local tests, feels a bit silly now that I wasn't checking the build logs properly..

File: torch/csrc/api/include/torch/nn/functional.h
Comment (gtamba) [reply torch/csrc/api/include/torch/nn/functional.h]: In general all C++ files should be padded with a newline? I think I got a bit confused on whether it applies everywhere or whether it follows a pattern

**Comment by gtamba** [comment]
@yf225  : Thank you so much for your patient reviewing!

**Review by yf225 — APPROVED** [review]
Thanks so much for the fantastic work @gtamba! I will wait for CI to pass and then merge it :D

**Comment by gtamba** [comment]
@yf225  : There's a failing test let me just look into that quickly

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: I suspect that we should change the options args here:
```suggestion
    torch::Tensor norm = F::normalize(input, NormalizeOptions().p(1).dim(-1));
```
which matches the Python version and should fixes the test errors :D

**Review by yf225 — CHANGES_REQUESTED** [review]
We would need to fix the test errors.

File: test/cpp/api/functional.cpp
Comment (gtamba) [reply test/cpp/api/functional.cpp]: Thanks for catching this, sorry for the messy workflow I'm building the source on a caveman laptop right now or I would have checked the tests locally before pushing them (working on changing that though)

**Comment by yf225** [comment]
@gtamba Thanks so much for the changes! I will iterate on some changes to try to fix the Windows build error :D

File: torch/csrc/api/include/torch/nn/options/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/normalization.h]: @gtamba I think we might need to add `#include <torch/nn/options/normalization.h>` in `torch/csrc/api/include/torch/nn/options.h` to fix the Windows build error. Thanks so much for your help!

**Review by yf225 — CHANGES_REQUESTED** [review]
@gtamba I think we might need to add `#include <torch/nn/options/normalization.h>` in `torch/csrc/api/include/torch/nn/options.h` to fix the Windows build error. Thanks so much for your help!

**Comment by gtamba** [comment]
@yf255 The windows build is still failing and I'm not sure what we're doing wrong here because of the ambiguous build logs :(

**Comment by yf225** [comment]
@gtamba No worries - I will look at it tomorrow. The Windows build is always weird :(

**Comment by yf225** [comment]
@gtamba I feel that we might need to add `TORCH_API` to all functions in `torch/csrc/api/include/torch/arg.h`, although it's really weird why we would need this. :/

File: torch/csrc/api/include/torch/nn/functional/normalization.h
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: When we confirm that `options = {}` doesn't cause the problem, I believe we should change it back to `options = {}`, and it would be awesome to have a test for the `F::normalize(input)` (i.e. no options) use case as well :D Thanks so much for your help @gtamba !

**Comment by gtamba** [comment]
> @gtamba I feel that we might need to add `TORCH_API` to all functions in `torch/csrc/api/include/torch/arg.h`, although it's really weird why we would need this. :/

I noticed that most other functions of the functional module also has a base Module and I couldn't find one which only has options included this way without a cpp counterpart declaration. I wonder if that could have something to do with it. But yeah I am confused from the error it's say that it can't resolve the Options struct's internal variables, but the Options struct is pretty much identical to any other options struct in functional/

**Comment by yf225** [comment]
@gtamba I think your assessment is correct, and I just pushed a commit to add `TORCH_API` to `arg.h` so that we don't need to have the cpp counterpart declaration. Hopefully that would work :)

**Review by yf225 — APPROVED** [review]
Thanks so much for the awesome work @gtamba!

File: torch/csrc/api/include/torch/nn/functional/normalization.h
Comment (gtamba) [reply torch/csrc/api/include/torch/nn/functional/normalization.h]: @yf225  : Can I still make these changes? I notice that facebook-bot is trying to land the pull request, and I didn't add the test for no options yet

Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/normalization.h]: @gtamba : Yes sorry this is my fault - didn't notice this before trying to land yesterday :( I saw you updated the PR (thanks a lot :D) and I will try to land it again

**Comment by yf225** [comment]
@gtamba Curious are you currently working on the Dropout modules? Please let me know if there is anything I can do to make the implementation easier. Thanks a lot for your help!

**Comment by gtamba** [comment]
@yf225  : Yes I'm currently working on the dropout modules, it took me a while to get started due to other commitments but I should have an initial PR up soon!

**Comment by yf225** [comment]
@gtamba Awesome thanks so much for your help! :D

**Comment by yf225** [comment]
@gtamba Awesome thanks so much for your help! Please keep me posted how it goes :D

**Comment by gtamba** [comment]
@yf225 : I am sorry for the delayed update, certain things came in the way and I could not work on this for a while. I noticed that #28424 has some relevant changes that might overlap with the rest of the dropout modules, I think I can follow the changes in DropOutOptions in this PR?

**Comment by yf225** [comment]
@gtamba : No worries about it at all! :D Yes I think https://github.com/pytorch/pytorch/pull/28424 is related and we also have some internal mandate to have the Dropout modules done within the next two weeks - curious do you mind if I take over the Dropout modules work instead? Thanks so much for your help again!

**Comment by gtamba** [comment]
@yf225 : That's totally fine it makes sense. If you're working on something else presently though, I should have a draft PR soon and it should wrap up shortly (+ code review) after the weekend, but iterations would make it risk taking too long. Feel free to hand me anything else. Again I feel sorry for the delay

**Comment by yf225** [comment]
@gtamba : Thanks so much for your help! Yes in that case I think if we can have it done over the next week or so it would be really awesome (I can personally help with the iterations by making changes to the PR if necessary). Please keep me posted on how it goes :D Thanks again!

**Pull request closed without merging**
